### PR TITLE
[#89] Add placement policy

### LINKED ISSUE #89: Container policy setting
Briefly touched in #47. There is a problem of container policy setting, `CreateBucket` doesn't have anything similar to this and `PutBucketPolicy` solves a bit different problem and implies that bucket already exists while container policy is set upon creation and can't be changed. I only see two possibilities at the moment:
 * misuse `LocationConstraint`, assign a set of magic predefined policies to each of locations and use them (which allows to create different containers, but looks weird)
 * store default policy in the session token (but it probably doesn't need that) or access box (which will make this access id/key combination always create containers with the same policy, but you can always have a number of ids/keys)
 * a combination of two is mapping location to policy in access box

Maybe there is something else we can do.

### PR DESCRIPTION
**Pull request by KirillovDenis**
closes #89 

Signed-off-by: Denis Kirillov <denis@nspcc.ru>

### CODE REVIEW

File: api/handler/put.go [hunk 1]
```diff
@@ -155,3 +173,22 @@ func parseBasicACL(basicACL string) (uint32, error) {
 		return uint32(value), nil
 	}
 }
+
+func (h *handler) writeError(w http.ResponseWriter, r *http.Request, msg, rid string, err error) {
```
Comment (masterSplinter01) [inline api/handler/put.go:177]: Isn't this function too similar to `func (h *handler) registerAndSendError(w http.ResponseWriter, r *http.Request, err error, logText string)`?

Comment (KirillovDenis) [reply api/handler/put.go]: You're right

**Review by masterSplinter01 — APPROVED** [review]

File: api/layer/layer.go [hunk 2]
```diff
@@ -498,11 +498,11 @@ func (n *layer) DeleteObjects(ctx context.Context, bucket string, objects []stri
 	return errs
 }
 
-func (n *layer) CreateBucket(ctx context.Context, p *CreateBucketParams) (*cid.ID, error) {
+func (n *layer) CreateBucket(ctx context.Context, p *CreateBucketParams, boxData *accessbox.Box) (*cid.ID, error) {
```
Comment (alexvanin) [inline api/layer/layer.go:501]: The same comment about `boxData` in `p`

**Review by alexvanin — APPROVED** [review]
LGTM

File: api/layer/container.go [hunk 3]
```diff
@@ -123,21 +123,14 @@ func (n *layer) containerList(ctx context.Context) ([]*BucketInfo, error) {
 	return list, nil
 }
 
-func (n *layer) createContainer(ctx context.Context, p *CreateBucketParams) (*cid.ID, error) {
+func (n *layer) createContainer(ctx context.Context, p *CreateBucketParams, boxData *accessbox.Box) (*cid.ID, error) {
```
[the remark this answers is not in the document]
Comment (KirillovDenis) [reply api/layer/container.go]: Fixed

File: api/layer/layer.go (hunk 2 above)
Comment (KirillovDenis) [reply api/layer/layer.go]: Fixed

File: api/handler/put.go [hunk 4]
```diff
@@ -90,43 +97,42 @@ func (h *handler) CreateBucketHandler(w http.ResponseWriter, r *http.Request) {
...
+		h.registerAndSendError(w, r, err, "could not parse body")
+		return
+	}
 
-		api.WriteErrorResponse(r.Context(), w, api.Error{
-			Code:           api.GetAPIError(api.ErrBadRequest).Code,
-			Description:    err.Error(),
-			HTTPStatusCode: http.StatusBadRequest,
-		}, r.URL)
+	boxData, err := getBoxData(r.Context())
```
Comment (alexvanin) [inline api/handler/put.go:110]: You can even 
```suggestion
	p.boxData, err = getBoxData(r.Context())
```

**Pull request merged**